### PR TITLE
Add the missing contrib documentation

### PR DESCRIPTION
**Pull request by pwaller**
* Document a bit about what's in the contrib directory.
* Add a manyclangs-setup.sh which installs prerequisites.

Signed-off-by: Peter Waller <p@pwaller.net>

### CODE REVIEW

File: contrib/manyclangs-setup.sh [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        (cd ninja-jobclient && python3 ./configure.py && ninja && cp ninja ${ELFSHAKER_BIN_DIR}/ninja-jobclient)
+    fi
+}
+
+install_ccache_from_source() {
+    if ! command -v ccache &> /dev/null
+    then
+        if [ ! -d ccache ]
+        then
+            git clone --depth 1 --single-branch --branch v4.6.1 https://github.com/ccache/ccache
```
Comment (veselink1) [inline contrib/manyclangs-setup.sh:174]: You mention that we use ccache v4.6.1 because it has some fixes in it. Any idea if those fixes are critical to building the packs? Because if they are, with our current detection strategy of `command -v ccache` we might not detect the case where an older ccache is already available in PATH.

File: contrib/manyclangs-setup.sh [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    fi
+    (cd llvm-project && git fetch origin)
+}
+
+install_jobserver() {
+    if ! command jobserver &> /dev/null
+    then
+        if [ ! -d jobclient ]
+        then
+            git clone https://github.com/olsner/jobclient
```
Comment (veselink1) [inline contrib/manyclangs-setup.sh:144]: Is it a good idea to checkout a specific commit here, for security reasons?

File: contrib/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+# Contrib scripts
+
+This directory anything which isn't exactly elfshaker but is nonetheless useful.
```
Comment (veselink1) [inline contrib/README.md:3]: ? This directory ~~anything which~~ isn't exactly elfshaker but is nonetheless useful.

File: contrib/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+The primary use case is to have one-big-machine and build a
+one-month-of-commits-packfile in one-go. Secondarily, we want to be able to keep
+the current month uptodate. Given this philosophy, we want to keep the
+one-big-machine as occupied as possible. Unfortunately there are various bits of
+the clang build process which are rather serial, resulting in a lot of unused
+CPU time.
+
+* To mitigate this, multiple builds run in parallel, but that conflicts with
+  making efficient use of incremental builds.
+* Therefore, builds need to be incremental, and it's necessary to multiple of
```
Comment (veselink1) [inline contrib/README.md:55]: There might be something wrong with this sentence. Doesn't appear to make much sense.

File: contrib/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  them run in parallel.
+  * Incremental builds start at a point in time and moves forward
+    commit-by-commit.
+    * This means a few tens-of-seconds build for a typical 'touch a few
```
Comment (veselink1) [inline contrib/README.md:59]: I know what you mean here but I found it a bit hard to parse. Suggestion:

"The incremental builds are fast, in the order of tens of seconds per build, and they can be run in parallel reasonably well because most of the work done for each build is not heavily parallelized."

**Review by veselink1 — APPROVED** [review]
LGTM with typos fixed and minor Qs answered, thanks. Let me know if you want me to fix any of these.

File: contrib/manyclangs-setup.sh (hunk 1 above)
Comment (pwaller) [reply contrib/manyclangs-setup.sh]: Good point, I think I will unconditionally build ccache for now, and pin it. I don't have a specific thing I can easily point to, except for experience. I expect this would be unnecessary at some point the future.

File: contrib/manyclangs-setup.sh (hunk 2 above)
Comment (pwaller) [reply contrib/manyclangs-setup.sh]: Yes, let's do that.

File: contrib/README.md (hunk 3 above)
Comment (pwaller) [reply contrib/README.md]: Rewritten:
> This directory contains scripts which make use of elfshaker, for example
scripts to build a [manyclangs](https://github.com/elfshaker/manyclangs) pack.

File: contrib/README.md (hunk 5 above)
Comment (pwaller) [reply contrib/README.md]: Rewritten much of the block:
>  * Incremental builds start at a point in time and move forward
    commit-by-commit.
>  * This results in a few tens-of-seconds build for a typical 'touch a few
    translation units' commit. However, such a build will only consume a few
    CPUs because they are bottlenecked on a small number (much fewer than
    `$NCPU`) of large translation units for that time.
>  * Saturating the machine to use all of the available resource is achieved by
    building from multiple points in time simultaneously.
>  * This introduces a problem of build contention, running more compiler
    procesess in parallel than necessary is wasteful and risks deadlocking the
    machine for example via an out-of-memory condition. To mitigate this, a
    jobserver-aware ninja is used, which ensure that only as many compilers run
    concurrently as there are available CPU cores.

**Comment by pwaller** [comment]
Thanks for spotting my mistakes, that was very helpful. I hope I've addressed your comments without introducing further mistakes. I've done another test of the setup script since my changes and all seems to work.

**Comment by pwaller** [comment]
Btw your comments remind me of John Reghr's blog post [You Might as Well Be a Great Copy Editor](https://blog.regehr.org/archives/1471). It's eye opening for me to see how many little mistakes crept in, although I'm not too surprised since I think I wrote the whole thing in about five minutes and didn't do much editing, so thanks for spending the time to help me iterate.

**Comment by veselink1** [comment]
Changes look good, thanks.
